File: algorithm/js/GenerateParenthesis.js
function generator(left, right, str, resultArray, n) {
    if(str.length == 2*n) {
        resultArray.push(str);
        return;
    }
    if(left < n) {
        generator(left + 1, right, str + "(", resultArray, n);
    }
    if((right < left)&&(right < n)) {
        generator(left, right + 1, str + ")", resultArray, n);
    }
}

/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    let resultArray = [];
    let str = "";

    generator(0, 0, str, resultArray, n);

    return resultArray;
};

File: algorithm/cpp/GenerateParenthesis.cpp
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        generate(0, 0, new StringBuilder(), n, result);
        return result;
    }

    public static void generate(int left, int right, StringBuilder sb, int n, List<String> result) {
        //若该位置可以填左括号，则尝试选填左括号
        if (left < n) {
            sb.append("(");
            left++;
            generate(left, right, sb, n, result);
            sb.deleteCharAt(sb.length() - 1);
            left--;
        }
        // 若已经填的位置中，左括号大于右括号数，则该位置可以填右括号
        if (right < left) {
            sb.append(")");
            right++;
            generate(left, right, sb, n, result);
            sb.deleteCharAt(sb.length() - 1);
            right--;
        }
        //如果所有的位置都已经填好，则符合要求的结果
        if (left == right && left == n) {
            result.add(sb.toString());
        }
    }
}
